src/Handler/TagHandler.php

```diff
@@ -117,9 +119,14 @@ public function addTags(array $tags)
      */
     public function invalidateTags(array $tags)
     {
-        $tagExpression = sprintf('(%s)(,.+)?$', implode('|', array_map('preg_quote', $this->escapeTags($tags))));
-        $headers = [$this->tagsHeader => $tagExpression];
-        $this->invalidator->invalidate($headers);
+        if (empty($tags)) {
+            return $this;
+        }
+
+        $this->invalidator->invalidateTags(
+            $this->escapeTags($tags),
```

Comment:
there is an issue here: the TagHandler does the escaping of the tags, but people could start calling CacheInvalidator::invalidateTags directly. at this point, the tag handler is down to a collection of tags. not sure if that is actually worth anything anymore. 

and actually, the escaping is again caching proxy specific. so instead, we should move escaping to the varnish client and ditch the tag handler completely i think. the added value of something that accumulates tags that we will invalidate is very small - one can just invalidate several times if needed, or build something to aggregate the tags.